### PR TITLE
Isolate Classpaths of CLI Runner and Servlet Container

### PR DESCRIPTION
**Pull request by f4lco**
**Problem solved:** this patch adds the classpath isolation we ought to have, and stops leaking Groovy, commons-cli, Logback, slf4j, onto the servlet container classpath.
This is crucially important, because Jetty 10 and Jetty 9 have different requirements towards Logback and slf4j. Prior to this patch, no matter which versions of Logback and slf4j we used, either Jetty 9 or Jetty 10 tests would fail our CI pipeline.
Consequently, this patch removes all exclusion lists, and the hacky `FilteringClassLoader`, which half-heartedly mitigated classpath leakage.

>  41 files changed, 204 insertions(+), 954 deletions(-)

This patch removes a fair bit of drag, was built on top of #268, and allows to green the build 🍏 

**Pending work:**
- [x] This patch likely breaks logging again, because the runner cannot easily reach into the class loader of the servlet container and reconfigure logging there. We won't have info/debug/error output of Gretty right now.
- [x] The "starter" code has a huge `// FIXME`. We will need to check to fill the gap how to get the servlet container classpath in starter code.

Please have a look again at the general concept:

https://github.com/gretty-gradle-plugin/gretty/blob/beff12a24d0da6f80441d03835f4f97733643987/libs/gretty-runner-jetty94/src/main/java/org/akhikhl/gretty/FilteringClassLoader.java#L89-L118

### CODE REVIEW

**Comment by boris-petrov** [comment]
I'm always glad when there's lots of code being removed. :smile: 

Great work! One big issue in Gretty is fixed with this. I understand the global picture, not so much the details but I believe you've done them right. :smile: I'm only not sure what you mean by `We will need to check to fill the gap how to get the servlet container classpath in starter code.`. Is that concerning the logging problem from your first point?

I've got a couple of minor style issues, otherwise thanks again for the great solution!

File: libs/gretty-core/src/main/groovy/org/akhikhl/gretty/ServerConfig.groovy [hunk 1]
```diff
@@ -20,7 +20,7 @@ import groovy.transform.TypeCheckingMode
 @ToString
 class ServerConfig {
 
-  static final int RANDOM_FREE_PORT = PortUtils.RANDOM_FREE_PORT
+  static final int RANDOM_FREE_PORT = -1
```
Comment (boris-petrov) [inline libs/gretty-core/src/main/groovy/org/akhikhl/gretty/ServerConfig.groovy:23]: You could reuse the `ServerDefaults.RANDOM_FREE_PORT` variable that you've added.

File: libs/gretty-runner-jetty10/src/main/groovy/org/akhikhl/gretty/JettyConfigurerImpl.groovy [hunk 2]
```diff
@@ -200,25 +199,12 @@ class JettyConfigurerImpl implements JettyConfigurer {
   def createWebAppContext(Map serverParams, Map webappParams) {
     List<String> webappClassPath = webappParams.webappClassPath
     JettyWebAppContext context = new JettyWebAppContext()
+    context.setThrowUnavailableOnStartupException(true)
```
Comment (boris-petrov) [inline libs/gretty-runner-jetty10/src/main/groovy/org/akhikhl/gretty/JettyConfigurerImpl.groovy:202]: Doesn't Groovy allow this to be written as `context.throwUnavailableOnStartupException = true`?

**Comment by ArLau** [comment]
Is there a possibility that the removal of the slf4j/ logback leakage will make in into gretty v3? This would be great as we cannot upgrade to gretty v4 (needing tomcat 9) and would like to update our used slf4j to v2.

**Comment by f4lco** [comment]
> Is there a possibility that the removal of the slf4j/ logback leakage will make in into gretty v3? This would be great as we cannot upgrade to gretty v4 (needing tomcat 9) and would like to update our used slf4j to v2.

Hello @ArLau, long time, no see 😜  Please send warm regards to everyone.
Yes, we intend to merge this patch into 3.x, fingers crossed. I'll probably ping you again once we have a release, so you guys can try it out & report back.

File: libs/gretty-core/src/main/groovy/org/akhikhl/gretty/ServerConfig.groovy (hunk 1 above)
Comment (f4lco) [reply libs/gretty-core/src/main/groovy/org/akhikhl/gretty/ServerConfig.groovy:23]: Unfortunately not, this constant moved to a shared module between the CLI runner and the servlet container runners. I do not want to drag in another module just for reusing one constant.
Besides that, now these two can also evolve independently. We might use -1 for the build DSL, and use a different sentinel value in the runner projects. A value of -1 is more of a historic fact than reasonable choice. Usually, binding the port 0 gives a 'random' free port.

File: libs/gretty-runner-jetty10/src/main/groovy/org/akhikhl/gretty/JettyConfigurerImpl.groovy (hunk 2 above)
Comment (f4lco) [reply libs/gretty-runner-jetty10/src/main/groovy/org/akhikhl/gretty/JettyConfigurerImpl.groovy:219]: Yes, there is no semantic difference between the two.

**Review by boris-petrov — APPROVED** [review]

**Pull request merged**